community/ndp/messaging-v1/src/main/java/org/neo4j/ndp/messaging/v1/PackStreamMessageFormatV1.java

```diff
@@ -0,0 +1,580 @@
+            case MessageTypes.MSG_IGNORED:     unpackIgnoredMessage( output ); break;
+            default:
+                throw new IOException( "Unknown message type: " + type );
+            }
+        }
+
+        @Override
+        public Reader reset( ReadableByteChannel channel )
+        {
+            this.unpacker.reset( channel );
+            return this;
+        }
+
+        private <E extends Exception> void unpackAckFailureMessage( MessageHandler<E> output )
+                throws E
+        {
+            output.handleAckFailureMessage();
+        }
+
+        private <E extends Exception> void unpackSuccessMessage( MessageHandler<E> output )
+                throws E, IOException
+        {
+            Map<String,Object> map = unpackRawMap();
+            output.handleSuccessMessage( map );
+        }
+
+        private <E extends Exception> void unpackFailureMessage( MessageHandler<E> output )
+                throws E, IOException
+        {
+            Map<String,Object> map = unpackRawMap();
+
+            String codeStr = (String) map.get( "code" );
+            String msg = (String) map.get( "message" );
+
+            output.handleFailureMessage( new Neo4jError( codeFromString( codeStr ), msg ) );
+        }
+
+        private <E extends Exception> void unpackIgnoredMessage( MessageHandler<E> output )
+                throws E
+        {
+            output.handleIgnoredMessage();
+        }
+
+        private <E extends Exception> void unpackItemMessage( MessageHandler<E> output )
+                throws E, IOException
+        {
+            long length = unpacker.unpackListHeader();
+            final Object[] fields = new Object[(int) length];
+            for ( int i = 0; i < length; i++ )
+            {
+                fields[i] = unpackValue();
+            }
+            output.handleRecordMessage( record( fields ) );
+        }
+
+        private <E extends Exception> void unpackRunMessage( MessageHandler<E> output )
+                throws E, IOException
+        {
+            String statement = unpacker.unpackString();
+            Map<String,Object> params = unpackRawMap();
+            output.handleRunMessage( statement, params );
+        }
+
+        private <E extends Exception> void unpackDiscardAllMessage( MessageHandler<E> output )
+                throws E, IOException
+        {
+            output.handleDiscardAllMessage();
+        }
+
+        private <E extends Exception> void unpackPullAllMessage( MessageHandler<E> output )
+                throws E, IOException
+        {
+            output.handlePullAllMessage();
+        }
+
+        private Map<String,Object> unpackRawMap() throws IOException
+        {
+            int size = (int) unpacker.unpackMapHeader();
+            if(size == 0)
+            {
+                return Collections.EMPTY_MAP;
+            }
+            Map<String, Object> map = new HashMap<>( size );
```

Comment:
Unfortunately this does not guarantee that map will not resize. This constructor uses default load factor which is 0.75 so resize might happen when map is 75% full. It is better to specify 1 as second param like this `new HashMap<>(size, 1)`. There are also some places where size is known but map's initial size is not specified. As this is serialization/deserialization I assume it might be important.